File: matlab-tools/HostPort/HostPort.m
classdef HostPort < handle
    %HOSTPORT HostPort class for communication to a microcontroller
    
    properties (SetAccess = private, Hidden)
        ptr_ = uint64(0);
    end
    
    properties (Dependent)
        Port (1,1) {mustBeInteger, mustBeNonnegative, mustBeLessThan(Port,256)}
        Baud (1,1) {mustBeInteger, mustBeNonnegative, mustBeLessThan(Baud,4294967296)}
        Header (1,1) {mustBeInteger, mustBeNonnegative, mustBeLessThan(Header,4294967296)}
        Terminator (1,1) {mustBeInteger, mustBeNonnegative, mustBeLessThan(Terminator,4294967296)}
        IsInit 
        %Pointer;
    end
    
    properties (SetAccess = private)
        Data 
    end
    
    methods(Static)
        
        function clear()
            HostPortMex('delete');
        end
        
        function ports = getPorts()
            ports = HostPortMex('getAvailablePort');
        end  
        
    end
    
    methods (Access = public)
        
        %constructor
        function obj = HostPort()
            %HOSTPORT Construct an instance of this classa
            obj.ptr_ = HostPortMex('new'); %assign pointer of new object
        end

        %destructor
        function delete(obj)
            if any(obj.ptr_ == HostPortMex('getHandles'))
                 HostPortMex('delete',obj.ptr_); %delete object
            end
        end

        %begin, restart and close mathods
        function exit = begin(obj, port, baud, header, terminator)
            %check input
            if rem(port,1)~=0 || port<0
                error('Port must be a positive integer');
            end
            if rem(baud,1)~=0 || baud<0
                error('Baud must be a positive integer');
            end
            if (nargin-1)<3
                exit = HostPortMex('begin',obj.ptr_,uint32(port),uint32(baud));
            else
                exit = HostPortMex('begin',obj.ptr_,uint32(port),uint32(baud),uint32(header),uint32(terminator));                
            end
        end

        function exit = restart(obj) 
            exit = HostPortMex('restart',obj.ptr_);
        end

        function exit = close(obj)
           exit = HostPortMex('close',obj.ptr_); 
        end
        
        %read and write
        function exit = read(obj, len)
           [obj.Data, exit] = HostPortMex('read',obj.ptr_,len); 
        end
        
        function exit = write(obj, buf)
            buf = typecast(buf, 'uint8');
            exit = HostPortMex('write',obj.ptr_,buf);            
        end
        
        function data = decode(obj, type)
            arguments
                obj
                type char {mustBeMember(type,{'uint8','uint16','uint32','single','double'})} = 'uint8'
            end
            data = typecast(obj.Data, type);
        end
        
        %logical operators
        function exit = logical(obj)
            exit = HostPortMex('isInit',obj.ptr_);
        end

        function exit = not(obj)
            exit = ~HostPortMex('isInit',obj.ptr_);
        end
        
    end

    methods %get & set methods
        
        function port = get.Port(obj)
            port = HostPortMex('getPort',obj.ptr_);
        end

        function baud = get.Baud(obj)
            baud = HostPortMex('getBaud',obj.ptr_);
        end

        function header = get.Header(obj)
            header = HostPortMex('getHeader',obj.ptr_);
        end

        function terminator = get.Terminator(obj)
            terminator = HostPortMex('getTerminator',obj.ptr_);
        end

        function isInit = get.IsInit(obj)
            isInit = HostPortMex('isInit',obj.ptr_);
        end

        %function ptr = get.Pointer(obj)
        %    ptr = obj.ptr_;
        %end

        function set.Port(obj, port)
           HostPortMex('setPort',obj.ptr_,uint32(port));
        end

        function set.Baud(obj, baud)
           if (baud < 0 || rem(baud,1)~=0)
               error('Baud must be a positive integer');
           end
           HostPortMex('setBaud',obj.ptr_,uint32(baud));
        end

        function set.Header(obj, header)
           if (header < 0 || rem(terminator,1)~=0)
               error('Header must be a header integer');
           end
           HostPortMex('setHeader',obj.ptr_,uint32(header)); 
        end

        function set.Terminator(obj, terminator)
           if (terminator < 0 || rem(terminator,1)~=0)
               error('Terminator must be a positive integer');
           end
           HostPortMex('setTerminator',obj.ptr_,uint32(terminator)); 
        end

    end
end

File: matlab-tools/HostPort/src/HostPort.cpp
/*
 * HostPort class to handle serial communication - source
 */

#include "HostPort.h"

//Begins
bool HostPort::begin(unsigned int port, unsigned int baud) {
    _terminator = TERMINATOR;
    _header = HEADER;
    _timeout = TIMEOUT;
    _port = port;
    _baud = baud;
    if (init(port, baud, _timeout)) {
        _isFirstRead = true;
        return true;
    }
    return false;
}

bool HostPort::begin(unsigned int port, unsigned int baud, unsigned int header, unsigned int terminator) {
    _terminator = terminator;
    _header = header;
    _timeout = TIMEOUT;
    _port = port;
    _baud = baud;
    if (init(port, baud, _timeout)) {
        _isFirstRead = true;
        return true;
    }
    return false;
}

bool HostPort::begin(unsigned int port, unsigned int baud, unsigned int header, unsigned int terminator, unsigned int timeout) {
    _terminator = terminator;
    _header = header;
    _port = port;
    _baud = baud;
    _timeout = timeout;
    if (init(port, baud, timeout)) {
        _isFirstRead = true;
        return true;
    }
    return false;
}

HostPort::~HostPort() {
    serial.~Serial();
}

HostPort::HostPort() {
    //empty
}

//set/get funs
bool HostPort::isInit(void) {
    return (bool)serial;
}
bool HostPort::setPort(unsigned int port) {
    //if (!serial) {
        _port = port;
        return true;
    //}
    //return false;
}
bool HostPort::setBaud(unsigned int baud) {
   // if (!serial) {
        _baud = baud;
        return true;
    //}
    //return false;
}
bool HostPort::setHeader(unsigned int header) {
    //if (!serial) {
        _header = header;
        return true;
    //}
    //return false;
}
bool HostPort::setTerminator(unsigned int terminator) {
    //if (!serial) {
        _terminator = terminator;
        return true;
    //}
    //return false;
}
unsigned int HostPort::getPort(void) {
    return _port;
}
unsigned int HostPort::getBaud(void) {
    return _baud;
}
unsigned int HostPort::getHeader(void) {
    return _header;
}
unsigned int HostPort::getTerminator(void) {
    return _terminator;
}

//reset & close fun
bool HostPort::close(void) {
    return serial.end();
}
bool HostPort::restart(void) {
    serial.end();
    return init(_port, _baud, _timeout);
}

//write
bool HostPort::write(unsigned char* packetPtr, unsigned int size) {

    if (!serial) {
        return false;
    }

    size_t c = 0; //counter for sent size

    //put header
    _tx_buf[c++] = _header & MASK;
    _tx_buf[c++] = (_header >> 8) & MASK;
    _tx_buf[c++] = (_header >> 16) & MASK;
    _tx_buf[c++] = (_header >> 24) & MASK;

    //add data
    for (int i = 0; i < size; ++i) _tx_buf[c++] = *(packetPtr + i);

    //put terminator
    _tx_buf[c++] = _terminator & MASK;
    _tx_buf[c++] = (_terminator >> 8) & MASK;
    _tx_buf[c++] = (_terminator >> 16) & MASK;
    _tx_buf[c++] = (_terminator >> 24) & MASK;

    //put start bytes in buf
    serial.write(_tx_buf, c);

    return true;

}

//read
bool HostPort::read(unsigned char* packetPtr, unsigned int size) {

    if (!serial) {
        return false;
    }

    if (_isFirstRead) {
        serial.flush();
        _isFirstRead = false;
    }

    unsigned int c = 0;
    unsigned char b;
    while (1) {
        if (c < 4) { //check header
            if (serial.read(&b, 1)>0) {
                if (b == ((_header >> (8 * c)) & MASK)) {
                    c++;
                    continue;
                }
            } else { 
                return false; 
            }
        } else if (c==4) { //read data
            if (serial.read(packetPtr, size) == 0) {
                return false; 
            };
            c++;
            continue;
        } else {
            if (serial.read(&b, 1) > 0) {
                if (b == ((_terminator >> (8 * (c - 5))) & MASK)) {
                    c++;
                    if (c == 9) { //all ok
                        return true;
                    }
                    continue;
                }
            } else { 
                break;
            }
        }
        break;
    }
    return false; //is ok???
}

//init
bool HostPort::init(unsigned int port, unsigned int baud, unsigned int timeout) {
    return serial.begin(port, baud, timeout);
}

